### ISSUE TITLE
No sanity checks on `destructure` and `loadparams!`

### ISSUE THREAD

**Issue by mcabbott** [FluxML/Flux.jl#1408]
Given too many parameters, or parameters of the wrong shapes, `destructure` and `loadparams!` silently have a go. I believe it would be safer to make these errors. Or at least warnings:
```julia
using Flux

m1 = Chain(Dense(2,2), Dense(2,2))
m1[1].W .= 1
v1, re1 = Flux.destructure(m1)
m2 = re1(rand(100)) # doesn't check length
m2[1].W # new values

m3 = Chain(Dense(2,2), Dense(2,2), Dense(2,2))
Flux.loadparams!(m1, params(m3)) # doesn't check length
m1[1].W # has been overwritten

m4 = Chain(Dense(2,2), Dense(2,3), Dense(3,2))
Flux.loadparams!(m1, params(m3)) # doesn't check shape either

Flux.loadparams!(m4, params(m1)) # ERROR: Expected param size (3, 2), got (2, 2)
```

**Comment by mcabbott** [comment]
> However, what if you wanted to `loadparams!` from a 3-layer MLP onto a 2-layer one or vice versa? Assuming the shape of the first two layers are the same, this should be supported in Flux. That is what `strict=False` in `load_state_dict` gives you: the ability to partially load params into a network without having to slice it up and reconstitute it afterwards (if indeed that's even possible in e.g. the presence of custom container layers).
> 
> _Originally posted by @ToucheSir in https://github.com/FluxML/Flux.jl/issues/1402#issuecomment-734959072_

The pytorch function is described for instance here (maybe note the best link?) https://pytorch.org/docs/master/generated/torch.nn.Module.html?highlight=load_state_dict#torch.nn.Module.load_state_dict

Perhaps we should follow that and make a keyword `strict=true` by default, which rejects all the size mismatches above. But what exactly is the rule for `strict=false`?

**Comment by darsnack** [comment]
If we want to address the `strict = false` case in a forward-thinking manner, then I think `params` needs to evolve past the "bag of arrays" approach to store structural information à la functors. This alone would make sanity checking better (by having a more narrowly scoped error case that `length(params(m)) != length(ps)`), and it would make educated-guessing for the `strict = false` case more reliable.

**Comment by mcabbott** [comment]
The obvious treelike functor-structure here is the model itself. Perhaps it is `loadparams!(m1, m5; strict=false)` which ought to work, with some rule like ignoring branches absent from either tree? Or does this have downsides I'm missing (like re-used parameter arrays?)

I know there's been discussion of re-designing Params, do you have a link to a summary / entry point on that?

**Comment by ToucheSir** [comment]
It would probably loklok something like https://github.com/FluxML/XLA.jl/blob/master/examples/conv.jl. in other words, Params may not be required at all.

**Comment by DhairyaLGandhi** [comment]
Correct, that's why you have https://github.com/FluxML/Functors.jl/pull/1 and Optimisers.jl, along with https://github.com/FluxML/Flux.jl/pull/1017 for training. 

This is the new api that we are moving towards.

**Comment by DhairyaLGandhi** [comment]
That's why I've updated Optimisers.jl recently to include most of the optimisers from flux (modulo some Adam derivatives but that is fairly easy)

**Comment by CarloLucibello** [comment]
Fixed — both now error rather than silently proceeding:

```
julia> re1(rand(100))
ERROR: DimensionMismatch: Rebuild expected a vector of length 12, got 100

julia> Flux.loadmodel!(Chain(Dense(2=>2)), Chain(Dense(2=>3)))
ERROR: DimensionMismatch: Tried to load size (3,) array into size (2,)
```

`loadparams!` has been replaced by `loadmodel!`, which walks the model structure and checks shapes, and `destructure` now lives in Optimisers.jl with a length check. Closing.
